pkg/verification/verification_service.go

```diff
@@ -87,3 +95,60 @@ func generateVerificationCode() (string, error) {
 
 	return string(buf), nil
 }
+
+// VerifyCode validates the user's phone verification code.  It updates the specified UserSignup value, so even
+// if an error is returned by this function the caller should still process changes to it
+func (s *ServiceImpl) VerifyCode(ctx *gin.Context, signup *v1alpha1.UserSignup, code string) error {
+
+	now := time.Now()
+
+	// If 24 hours has passed since the verification timestamp, then reset the timestamp and verification attempts
+	ts, err := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey])
+	if err != nil || (err == nil && now.After(ts.Add(24*time.Hour))) {
+		// Set a new timestamp
+		signup.Annotations[v1alpha1.UserSignupVerificationTimestampAnnotationKey] = now.Format(TimestampLayout)
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(0)
+	}
+
+	attemptsMade, err := strconv.Atoi(signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey])
+	if err != nil {
+		// We shouldn't get an error here, but if we do, we should probably set verification attempts to max allowed
+		// so that we at least now have a valid value
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(s.config.GetVerificationAttemptsAllowed())
+		return err
+	}
+
+	// If the user has made more attempts than is allowed per day, return an error
+	if attemptsMade >= s.config.GetVerificationAttemptsAllowed() {
+		return errors2.NewTooManyRequestsError("too many verification attempts")
+	}
+
+	exp, err := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey])
+	if err != nil {
+		// If the verification expiry timestamp is corrupt or missing, then return an error
+		return errors2.NewInternalError(err)
+	}
+
+	if now.After(exp) {
+		// If it is now past the expiry timestamp for the verification code, return a 400 Bad Request error
+		return errors2.NewBadRequest("verification code expired")
+	}
+
+	// If the code matches then set VerificationRequired to false, reset other verification annotations
+	if code == signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] {
+		signup.Spec.VerificationRequired = false
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCodeAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationAttemptsAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationCounterAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserSignupVerificationTimestampAnnotationKey)
+		delete(signup.Annotations, v1alpha1.UserVerificationExpiryAnnotationKey)
+		return nil
+	} else {
+		// If the code doesn't match
+		attemptsMade++
+		signup.Annotations[v1alpha1.UserVerificationAttemptsAnnotationKey] = strconv.Itoa(attemptsMade)
+		return errors2.NewBadRequest("invalid code")
```

Comment:
i think bad request more means that the request couldn't be understood: 

```
The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.

https://www.restapitutorial.com/httpstatuscodes.html#
```

in this case, it's more that the user is unauthorized, so maybe we should send back 401 unauthorized?

Reply:
> in this case, it's more that the user is unauthorized, so maybe we should send back 401 unauthorized?

We should be careful with 401 errors. The frontend could have a global routing for all 401 errors and redirect users to the login page.
Maybe 403 would be better than 401 here?

Reply:
hmm 403 may be a bit misleading: 

```
The server understood the request, but is refusing to fulfill it. Authorization will not help and the request SHOULD NOT be repeated. 
```

is there anyway for us to confirm if there is a 401 catch-all? i guess if we have no other options that's the best we can do..

Reply:
I still have a feeling that 401 can be even more misleading. This endpoint requires a user token. So, the user is authenticated correctly. We could confirm with RHD folks if they use any 401 redirects but we don't know if RHD is the only UI which will use our reg-service backend. We had this issue with 401 redirects previously with OSIO.

Reply:
hmm yeah, 401 may be confusing because the user may not understand whether it's the token or the phone verification that failed and if we know its been an issue in the past than maybe 403 + error message <phone verification token invalid/expired/etc> will help make it clear..

Reply:
I struggled to find a suitable response code for this case.  Bad Request was the best I could really come up with, however I'm still open to alternate suggestions.

Reply:
On further reflection I do think that 403 is a better option than 400.  I'll update the code to that for now.